pkgs/stdenv/generic/check-meta.nix

```diff
@@ -372,7 +444,56 @@ let
         }" ]
     else
       [ "key 'meta.${k}' is unrecognized; expected one of: \n  [${concatMapStringsSep ", " (x: "'${x}'") (attrNames metaTypes)}]" ];
-  checkMeta = meta: optionals config.checkMeta (concatMap (attr: checkMetaAttr attr meta.${attr}) (attrNames meta));
+  checkMeta = meta: optionals config.checkMeta (concatMap (attr: checkMetaAttr attr meta.${attr}) (attrNames meta)
+
+    # Extended checks for problems, as they do not fit the module system
+    ++ lib.optionals (meta ? problems) (
+
+      # Check problem kinds are correct
+      lib.pipe meta.problems [
+        (lib.mapAttrs (name: problem: { kind = name; } // problem))
+        (lib.mapAttrsToList (
+          name: problem:
+          lib.mapAttrsToList (checkMetaAttr "meta.problems.${name}" meta.problemTypes) problem
+        ))
+        (lib.concatLists)
+        (lib.remove null)
+      ]
+
+      # Check that problem has a message (required field)
+      ++ lib.pipe meta.problems [
+        (lib.mapAttrsToList (
+          name: problem:
+          if problem ? message then
+            null
+          else
+            "key 'meta.problems.${name}.message' is missing"
+        ))
+        (lib.remove null)
+      ]
+
+      # Check that some problem kinds are unqiue
+      ++ lib.pipe meta.problems [
+        (lib.mapAttrs (name: problem: { kind = name; } // problem))
+        # Count the number of instances of each problem kind,
```

Comment:
I think the manual counting & checking is better implemented by combining `groupBy` & `concatMap` over `problemKindsUnique'`.